internal/recursive_delete.go

```diff
@@ -0,0 +1,66 @@
+package internal
+
+import (
+	"log"
+	"os"
+	"path/filepath"
+)
+
+type fsNode struct {
+	Parent   *fsNode
+	Children map[string]*fsNode
+	HasFiles bool
+	Path     string
+}
+
+// Recurse to the leaf directories looking for empty directories to delete
+// and then work our way back up deleting everything that becomes empy during
+// the process.
+func deleteEmptyDirs(n *fsNode) error {
+	f, err := os.Open(n.Path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	dirs, err := f.ReadDir(-1)
+	if err != nil {
+		return err
+	}
+	for _, d := range dirs {
+		if d.IsDir() {
+			child := &fsNode{
+				Parent:   n,
+				Children: make(map[string]*fsNode),
+				HasFiles: false,
+				Path:     filepath.Join(n.Path, d.Name()),
+			}
+			n.Children[child.Path] = child
+			deleteEmptyDirs(child)
+		} else {
+			n.HasFiles = true
+		}
+	}
+
+	if n.Parent != nil && !n.HasFiles && len(n.Children) == 0 {
+		if err := os.Remove(n.Path); err != nil {
+			// We can only really warn here. People are going do do stuff like:
+			// docker run -v /run/secrets/foo:/foo
+			// which could prevent us from being able to delete the directory
+			log.Printf("Unable to remove empty directory: %s - %s", n.Path, err)
+		} else {
+			delete(n.Parent.Children, n.Path)
+		}
+	}
+	return nil
+}
+
+func DeleteEmptyDirs(path string) error {
+	top := &fsNode{
+		Parent:   nil,
+		Children: make(map[string]*fsNode),
+		HasFiles: false,
+		Path:     path,
+	}
+	return deleteEmptyDirs(top)
+}
```

Comment:
It looks like you put everything right above, it just looks like it could be done easier:
- as you already leverage the recursion - there is no need to pass state back and forth;
- it should be possible to make it tail-recursive to allow the compiler optimizer to unwrap it into a loop.

I tried to turn the above into this snippet:
```Golang
func DeleteEmptySubdirs(path string) bool {
  // returns True if and only if the path is an empty directory
  // deletes all empty sub-directories
  node, err := os.Open(path);
  if err != nil { warnAndReturnFalse }
  defer node.Close()
  children, err := node.ReadDir(-1)
  if err != nil { warnAndReturnFalse }
  isEmpty := true
  for _, child := range children {
    if child.IsDir() {
      childPath := filepath.Join(path, child.Name())
      if DeleteEmptySubdirs(childPath) {
        if err := os.Remove(childPath); err != nil {warnAndReturnFalse}
      } else {
        isEmpty = false
      }
    } else {
      isEmpty = false
    }
  }
  return isEmpty;
}
```

Reply:
i was trying to avoid (ab)using bool or error for handling 2 different things. but you did give me an idea where i can keep an awkward private function and wrap it with a clean public function:

c6278c9c61f3bad4215f0e2bd4381e0a5b126c15

Reply:
nice, thanks